module-system/module-implementations/sov-nft-module/src/offchain.rs

```diff
@@ -0,0 +1,142 @@
+    let collection_address = nft.get_collection_address().to_string();
+    let nft_id = nft.get_token_id();
+    let new_owner_str = nft.get_owner().to_string();
+    let frozen = nft.is_frozen();
+    let metadata_url = nft.get_token_uri();
+    let old_owner_address = old_owner.map(|x| x.to_string());
+
+    tokio::task::block_in_place(|| {
+        if let Ok(conn_string) = std::env::var("POSTGRES_CONNECTION_STRING") {
+            let mut client = postgres::Client::connect(&conn_string, NoTls).unwrap();
+
+            // Check current owner in the database for the NFT
+            let rows = client
+                .query(
+                    "SELECT owner FROM nfts WHERE collection_address = $1 AND nft_id = $2",
+                    &[&collection_address, &(nft_id as i64)],
+                )
+                .unwrap();
+
+            let db_owner: Option<String> = rows.get(0).map(|row| row.get(0));
+
+            // Handle ownership change logic for top_owners table
+            if let Some(db_owner_str) = db_owner {
+                if old_owner_address.is_none() {
+                    // This means it's a mint operation but the NFT already exists in the table.
+                    // Do nothing as we shouldn't increment in this scenario.
+                } else if old_owner_address.as_ref() != Some(&new_owner_str) {
+                    // Transfer occurred
+
+                    // Decrement count for the database owner (which would be the old owner in a transfer scenario)
+                    let _ = client.execute(
+                        "UPDATE top_owners SET count = count - 1 \
+                        WHERE owner = $1 AND collection_address = $2 AND count > 0",
+                        &[&db_owner_str, &collection_address],
+                    );
+
+                    // Increment count for new owner
+                    let _ = client.execute(
+                        "INSERT INTO top_owners (owner, collection_address, count) VALUES ($1, $2, 1) \
+                        ON CONFLICT (owner, collection_address) \
+                        DO UPDATE SET count = top_owners.count + 1",
+                        &[&new_owner_str, &collection_address],
+                    );
+                }
+            } else if old_owner_address.is_none() {
+                // Mint operation, and NFT doesn't exist in the database. Increment for the new owner.
+                let _ = client.execute(
+                    "INSERT INTO top_owners (owner, collection_address, count) VALUES ($1, $2, 1) \
+                    ON CONFLICT (owner, collection_address) \
+                    DO UPDATE SET count = top_owners.count + 1",
+                    &[&new_owner_str, &collection_address],
+                );
+            }
+
+            // Update NFT information after handling top_owners logic
+            let _ = client.execute(
+                "INSERT INTO nfts (\
```

Comment:
What about extracting all the sql strings as constants in a separate `sql.rs` file?